cmd/endpointQuerier/main.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"context"
+	"encoding/xml"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptrace"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"../../internal/endpoints"
+	"../../internal/fhir"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Metrics collected inside build-in prometheus vector. Each METRIC has its own registrations
+var httpCodesGaugeVec *prometheus.GaugeVec
+var responseTimeGaugeVec *prometheus.GaugeVec
+var tlsVersionGaugeVec *prometheus.GaugeVec
+var fhirVersionGaugeVec *prometheus.GaugeVec
+var totalUptimeChecksCounterVec *prometheus.CounterVec
+var totalFailedUptimeChecksCounterVec *prometheus.CounterVec
+
+// Need to define timeout or else it is infinite
+var netClient = &http.Client{
+	Timeout: time.Second * 35,
+}
+
+func getHTTPRequestTimingFor(url string, organizationName string, recordLongRunning bool) {
+	// Specifically query the FHIR endpoint metadata
+	url += "metadata"
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		println("HTTP Request Error: %s", err)
+	}
+
+	var start time.Time
+	trace := &httptrace.ClientTrace{}
+
+	// Drop connection if no reply within 30 seconds
+	ctx, cancel := context.WithDeadline(req.Context(), time.Now().Add(30*time.Second))
+	// Cancel the context once we are done with this function so that context does not remain in memory (causing a leak)
+	defer cancel()
+	req = req.WithContext(httptrace.WithClientTrace(ctx, trace))
```

Comment:
https://golang.org/pkg/context/  My understanding is that contexts allow you to better handle calls outside of your control (in this case setting a deadline so we can cancel it if it runs past a deadline and also allows you to cancel a request even if it is running in a different process. If you run a call in a different process, so long as you pass a context along with it you'll still be able to cancel it because it is running in the context you defined.